pkg/auth/handler/mfa/create_oob.go

```diff
@@ -0,0 +1,156 @@
+package mfa
+
+import (
+	"net/http"
+
+	"github.com/skygeario/skygear-server/pkg/auth"
+	"github.com/skygeario/skygear-server/pkg/auth/dependency/mfa"
+	coreAuth "github.com/skygeario/skygear-server/pkg/core/auth"
+	"github.com/skygeario/skygear-server/pkg/core/auth/authz"
+	"github.com/skygeario/skygear-server/pkg/core/auth/authz/policy"
+	"github.com/skygeario/skygear-server/pkg/core/config"
+	"github.com/skygeario/skygear-server/pkg/core/db"
+	"github.com/skygeario/skygear-server/pkg/core/handler"
+	"github.com/skygeario/skygear-server/pkg/core/inject"
+	"github.com/skygeario/skygear-server/pkg/core/mail"
+	"github.com/skygeario/skygear-server/pkg/core/phone"
+	"github.com/skygeario/skygear-server/pkg/core/server"
+	"github.com/skygeario/skygear-server/pkg/core/skyerr"
+)
+
+func AttachCreateOOBHandler(
+	server *server.Server,
+	authDependency auth.DependencyMap,
+) *server.Server {
+	server.Handle("/mfa/oob/new", &CreateOOBHandlerFactory{
+		Dependency: authDependency,
+	}).Methods("OPTIONS", "POST")
+	return server
+}
+
+type CreateOOBHandlerFactory struct {
+	Dependency auth.DependencyMap
+}
+
+func (f CreateOOBHandlerFactory) NewHandler(request *http.Request) http.Handler {
+	h := &CreateOOBHandler{}
+	inject.DefaultRequestInject(h, f.Dependency, request)
+	return handler.RequireAuthz(handler.APIHandlerToHandler(h, h.TxContext), h.AuthContext, h)
+}
+
+func (h *CreateOOBHandler) ProvideAuthzPolicy() authz.Policy {
+	return policy.AllOf(
+		authz.PolicyFunc(policy.DenyNoAccessKey),
+		authz.PolicyFunc(policy.RequireAuthenticated),
+		authz.PolicyFunc(policy.DenyDisabledUser),
+	)
+}
+
+type CreateOOBRequest struct {
+	Channel coreAuth.AuthenticatorOOBChannel `json:"channel"`
+	Phone   string                           `json:"phone"`
+	Email   string                           `json:"email"`
+}
+
+func (r CreateOOBRequest) Validate() error {
+	switch r.Channel {
+	case coreAuth.AuthenticatorOOBChannelSMS:
+		return phone.Parse(r.Phone)
+	case coreAuth.AuthenticatorOOBChannelEmail:
+		return mail.Parse(r.Email)
+	default:
+		return skyerr.NewInvalidArgument("invalid channel", []string{"channel"})
+	}
+}
+
+type CreateOOBResponse struct {
+	AuthenticatorID   string `json:"authenticator_id"`
+	AuthenticatorType string `json:"authenticator_type"`
+	Channel           string `json:"channel"`
+}
+
+// @JSONSchema
+const CreateOOBRequestSchema = `
+{
+	"$id": "#CreateOOBRequest",
+	"oneOf": [
```

Comment:
Just want to note that the presence of `phone` or `email` must depend on the `channel`. i.e. `additionalProperties` should be false.